hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -3146,3 +3173,70 @@ func globalConfigString(hcluster *hyperv1.HostedCluster) (string, error) {
 	return supportutil.HashSimple(config + targetVersion + pullSecretName + globalConfig)
 }
+
+// mirrorConfigs mirrors higher-level machineConfiguration objects, i.e., kubeletConfig and containerRuntimeConfig, into
+// the HCP namespace, since those are needed as an input for certain operators (such as NTO)
+func (r *NodePoolReconciler) mirrorConfigs(ctx context.Context, logr logr.Logger, mirroredConfigs []corev1.ConfigMap, controlPlaneNamespace string, nodePool *hyperv1.NodePool) error {
+	// get kubeletConfig objects from the HCP namespace
+	kubeletConfigConfigMapList := &corev1.ConfigMapList{}
+	if err := r.List(ctx, kubeletConfigConfigMapList, &client.ListOptions{
+		Namespace:     controlPlaneNamespace,
+		LabelSelector: labels.SelectorFromValidatedSet(labels.Set{mirroredConfigLabel: ""}),
+	}); err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	// get containerRuntimeConfig objects from the HCP namespace
+	containerRuntimeConfigConfigMapList := &corev1.ConfigMapList{}
+	if err := r.List(ctx, containerRuntimeConfigConfigMapList, &client.ListOptions{
+		Namespace:     controlPlaneNamespace,
+		LabelSelector: labels.SelectorFromValidatedSet(labels.Set{mirroredConfigLabel: ""}),
+	}); err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	existingConfigs := append(kubeletConfigConfigMapList.Items, containerRuntimeConfigConfigMapList.Items...)
+
+	// need to delete the redundant configs, since they are no longer part of the nodepool spec
+	if len(existingConfigs) > len(mirroredConfigs) {
+		var configForDeletion []*corev1.ConfigMap
+		for i := 0; i < len(existingConfigs); i++ {
+			var found bool
+			existingConfig := &existingConfigs[i]
+			for _, mirroredConfig := range mirroredConfigs {
+				// the object was found in the node pool NS - keep it.
+				if existingConfig.Name == composeConfigName(mirroredConfig.Name, nodePool.Name) {
+					found = true
+					break
+				}
+			}
+			if !found {
+				configForDeletion = append(configForDeletion, existingConfig)
+			}
+		}
+		for _, config := range configForDeletion {
+			_, err := supportutil.DeleteIfNeeded(ctx, r.Client, config)
+			if err != nil {
+				return fmt.Errorf("failed to delete ConfigMap %s/%s: %w", config.Namespace, config.Name, err)
+			}
+		}
+	}
+	// update or create the configs that need to be mirrored into the HCP NS
+	for _, config := range mirroredConfigs {
+		cm := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      composeConfigName(config.Name, nodePool.Name),
+				Namespace: controlPlaneNamespace},
+		}
+		if result, err := r.CreateOrUpdate(ctx, r.Client, cm, func() error {
+			return reconcileMirroredConfigs(cm, &config, nodePool)
+		}); err != nil {
+			return fmt.Errorf("failed to reconcile %s/%s ConfigMap: %w", config.Namespace, config.Name, err)
+		} else {
+			logr.Info("Reconciled ConfigMap", "kind", config.Kind, "result", result)
+		}
+	}
+	return nil
+}
+
+func composeConfigName(configName, nodepoolName string) string {
```

Comment:
I believe we talked about this in another PR but please apply the correct logic to names. Both the `configName` and `nodepoolName` may be at the length limit for a k8s identifier. You must emit a composed name here that is a valid k8s identifier.

Reply:
Yes. I'll be using the function from the second PR.